### PR TITLE
Add DLND to moab coupler

### PR DESCRIPTION
**Pull request by iulian787**
Add data land capability for moab driver;
case tested: --res r05_r05 --compset RMOSGPCC 
standard data moab model, mainly adapt from data ocean implementation

major change: Initialize all factors set by MOAB area correction init routine to 1.0. It affects all models, 
but it seems that some values were not initialized for data land case.
On chrysalis, with intel compiler, it actually complained about non initialized values when applying correction factors. 

The compressing file errors seen during development were caused by non-ascii characters in the log files. 
Those were caused by an uninitialized mapper, for which we were still creating a MOAB map. It was never used, 
it is fixed properly now, the moab map is not computed anymore. 

[BFB]

### CODE REVIEW

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 [hunk 1]
```diff
@@ -135,6 +145,18 @@ subroutine dlnd_comp_init(Eclock, x2l, l2x, &
     character(nec_len) :: nec_str   ! elevation class, as character string
     character(*), parameter :: domain_fracname_unset = 'null'
 
+#ifdef HAVE_MOAB
+    character*400  tagname
+    real(R8) latv, lonv
+    integer iv, tagindex, ilat, ilon  !, arrsize, nfields ! ierr is already defined as integer above
```
Comment (vijaysm) [inline components/data_comps/dlnd/src/dlnd_comp_mod.F90:151]: Remove commmented out fields

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 [hunk 2]
```diff
@@ -135,6 +145,18 @@ subroutine dlnd_comp_init(Eclock, x2l, l2x, &
...
+    real(R8), allocatable, target :: data(:)
+    integer(IN), pointer :: idata(:)   ! temporary
+    real(r8), dimension(:), allocatable :: moab_vert_coords  ! temporary
```
Comment (vijaysm) [inline components/data_comps/dlnd/src/dlnd_comp_mod.F90:154]: Use R8 or r8 consistently everywhere. I suggest R8

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 [hunk 3]
```diff
@@ -256,6 +278,119 @@ subroutine dlnd_comp_init(Eclock, x2l, l2x, &
 
     call t_stopf('dlnd_initmctdom')
 
+#ifdef HAVE_MOAB
+    ilat = mct_aVect_indexRA(ggrid%data,'lat')
+    ilon = mct_aVect_indexRA(ggrid%data,'lon')
+    allocate(moab_vert_coords(lsize*3))
+    do iv = 1, lsize
+       lonv = ggrid%data%rAttr(ilon, iv) * SHR_CONST_PI/180.
+       latv = ggrid%data%rAttr(ilat, iv) * SHR_CONST_PI/180.
```
Comment (vijaysm) [inline components/data_comps/dlnd/src/dlnd_comp_mod.F90:287]: We should expose a utility method in MOAB to do coordinate transformations. It doesn't make sense having this exact same type of transform code in every component

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 [hunk 4]
```diff
@@ -256,6 +278,119 @@ subroutine dlnd_comp_init(Eclock, x2l, l2x, &
...
+                                     idata)
+    if (ierr .ne. 0)  &
+       call shr_sys_abort('Error: fail to set GLOBAL_ID tag ')
+ 
+    ierr = iMOAB_ResolveSharedEntities( mlnid, lsize, idata );
+    if (ierr .ne. 0)  &
+       call shr_sys_abort('Error: fail to resolve shared entities')
+ 
+    deallocate(moab_vert_coords)
+    deallocate(idata)
```
Comment (vijaysm) [inline components/data_comps/dlnd/src/dlnd_comp_mod.F90:320]: Is idata allocated inside MCT?

File: components/data_comps/dlnd/src/lnd_comp_mct.F90 [hunk 5]
```diff
@@ -146,13 +152,25 @@ subroutine lnd_init_mct( EClock, cdata, x2l, l2x, NLFilename )
...
+#endif
     call dlnd_comp_init(Eclock, x2l, l2x, &
          seq_flds_x2l_fields, seq_flds_l2x_fields, &
          SDLND, gsmap, ggrid, mpicom, compid, my_task, master_task, &
          inst_suffix, inst_name, logunit, read_restart, &
          scmMode, scmlat, scmlon)
-
+#ifdef HAVE_MOAB
+     if (my_task == master_task) then
+          call seq_infodata_PutData( infodata, lnd_domain=SDLND%domainFile) ! we use the same one for regular case
```
Comment (vijaysm) [inline components/data_comps/dlnd/src/lnd_comp_mct.F90:169]: Why is this needed only for MOAB driver?

File: driver-moab/main/component_mod.F90 [hunk 6]
```diff
@@ -748,6 +748,7 @@ subroutine component_init_areacor_moab (comp, mbccid, mbcxid, seq_flds_c2x_fluxe
          lsize = comp(1)%mblsize
          allocate(areas (lsize, 3)) ! lsize is along grid; read mask too
          allocate(factors (lsize, 2))
+         factors = 1.0 ! initialize with 1.0 all factors; then maybe correct them
```
Comment (vijaysm) [inline driver-moab/main/component_mod.F90:751]: Shouldn't it be factors(:,:) = 1.0? Or is the syntax in general the same (and independent of dimension or compiler family)?

**Review by vijaysm — COMMENTED** [review]
Reviewed the code and made comments. I haven't run this branch and tested it just yet.

**Comment by iulian787** [comment]
I tested this branch on different platforms, and chrysalis is working fine; gce, perlmutter, and my laptop (gnu compiler) all fail to compress the log files at the end. So the bug I saw earlier is not solved yet. Somehow, it worked on gce when I tested 2 weeks ago;  needs more work

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 (hunk 3 above)
Comment (iulian787) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:287]: We have this in MOAB, but it does not make sense to expose it from there; iMOAB?  The utility should be either in 
E3SM/share or in E3SM/driver-moab/shr

File: components/data_comps/dlnd/src/lnd_comp_mct.F90 (hunk 5 above)
Comment (iulian787) [reply components/data_comps/dlnd/src/lnd_comp_mct.F90:169]: the regular coupler does not need the mesh; we do

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 (hunk 4 above)
Comment (iulian787) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:320]: mct_gsMap_orderedPoints  allocates it

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 (hunk 3 above)
Comment (rljacob) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:287]: If its only used within the coupler code (including the code in the models) then driver-moab/shr would be the place.

Comment (rljacob) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:287]: But you don't have to do that for this PR.

File: driver-moab/main/component_mod.F90 (hunk 6 above)
Comment (iulian787) [reply driver-moab/main/component_mod.F90:751]: we are not consistent, we use both forms, but this one is shorter. 
We are consistent when we use set/get methods with these 2 dimensional arrays, iMOAB_SetDoubleTagStorage, iMOAB_GetDoubleTagStorage; 
`ierr = iMOAB_SetDoubleTagStorage( mbccid, tagname, arrsize , comp(1)%mbGridType, factors)`

File: driver-moab/main/prep_lnd_mod.F90 [hunk 7]
```diff
@@ -693,7 +693,7 @@ subroutine prep_lnd_mrg_moab (infodata)
 #endif
 #ifdef MOABCOMP
     character(CXX)           :: tagname, mct_field
-    real(r8)                 :: difference
+    real(R8)                 :: difference
```
[the remark this answers is not in the document]
Comment (rljacob) [reply driver-moab/main/prep_lnd_mod.F90:696]: Because you said to use R8 or r8 consistently.  R8 is better.

File: components/data_comps/dlnd/src/dlnd_comp_mod.F90 [hunk 8]
```diff
@@ -256,6 +278,119 @@ subroutine dlnd_comp_init(Eclock, x2l, l2x, &
...
+    enddo
+ 
+    ! create the vertices with coordinates from MCT domain
+    ierr = iMOAB_CreateVertices(mlnid, lsize*3, 3, moab_vert_coords)
+    if (ierr .ne. 0)  &
+       call shr_sys_abort('Error: fail to create MOAB vertices in data lnd model')
+ 
+    tagname='GLOBAL_ID'//C_NULL_CHAR
+    ierr = iMOAB_DefineTagStorage(mlnid, tagname, &
+                                  0, & ! dense, integer
```
[the remark this answers is not in the document]
Comment (iulian787) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:300]: yes, we do, because if we do not define it, iMOAB set tag does not work

[the remark this answers is not in the document]
Comment (vijaysm) [reply components/data_comps/dlnd/src/dlnd_comp_mod.F90:303]: Hmm, it should for a fundamental tag like this.

**Review by vijaysm — APPROVED** [review]
Thanks for addressing my comments!
